HopsanGUI/Widgets/TextEditorWidget.cpp

```diff
@@ -0,0 +1,682 @@
+            textCursor().beginEditBlock();
+            QString text = textCursor().selection().toPlainText();
+            text.replace("\n// ", "\n");
+            if(text.startsWith("// ")) {
+                text.remove(0,3);
+            }
+            int pos = textCursor().position();
+            int stopPos = textCursor().anchor();
+            QTextCursor c;
+            c = textCursor();
+            textCursor().removeSelectedText();
+            insertPlainText(text);
+
+            c.setPosition(std::min(pos,stopPos));
+            setTextCursor(c);
+
+            textCursor().endEditBlock();
+        }
+        else if(mLanguage == HighlighterTypeEnum::Hcom) {
+            textCursor().beginEditBlock();
+            QString text = textCursor().selection().toPlainText();
+            text.replace("\n# ", "\n");
+            if(text.startsWith("# ")) {
+                text.remove(0,2);
+            }
+            int pos = textCursor().position();
+            int stopPos = textCursor().anchor();
+            QTextCursor c;
+            c = textCursor();
+            textCursor().removeSelectedText();
+            insertPlainText(text);
+
+            c.setPosition(std::min(pos,stopPos));
+            setTextCursor(c);
+
+            textCursor().endEditBlock();
+        }
+    }
+    else
+    {
+        QPlainTextEdit::keyPressEvent(event);
+
+        updateAutoCompleteList();
+
+        QTextCursor tc = textCursor();
+        tc.select(QTextCursor::WordUnderCursor);
+        QString prefix = tc.selectedText();
+        if((event->key() != Qt::Key_Space || !event->modifiers().testFlag(Qt::ControlModifier)) && prefix.isEmpty())
+            return;
+
+        if (prefix != mpCompleter->completionPrefix()) {
+            mpCompleter->setCompletionPrefix(prefix);
+            mpCompleter->popup()->setCurrentIndex(mpCompleter->completionModel()->index(0, 0));
+        }
+        QRect cr = cursorRect();
+        cr.setWidth(mpCompleter->popup()->sizeHintForColumn(0)
+                    + mpCompleter->popup()->verticalScrollBar()->sizeHint().width());
+        mpCompleter->complete(cr);
+    }
+}
+
+
+void TextEditor::wheelEvent(QWheelEvent* event)
+{
+#if QT_VERSION >= 0x050000  //zoomIn() and zoomOut() not available in Qt4
+   if ((event->modifiers() == Qt::ControlModifier) && (event->delta() > 0))
+       zoomIn(2);
+   else if ((event->modifiers() == Qt::ControlModifier) && (event->delta() < 0))
+       zoomOut(2);
+   else
+       QPlainTextEdit::wheelEvent(event);
+#else
+   QPlainTextEdit::wheelEvent(event);
+#endif
+}
+
+
+//! @brief Updates list of auto complete words
+void TextEditor::updateAutoCompleteList()
+{
+    if(mLanguage == HighlighterTypeEnum::Hcom) {
+        mpCompleter->setModel(new QStringListModel(gpTerminalWidget->mpHandler->getAutoCompleteWords(), mpCompleter));
+    }
+    else if(mLanguage == HighlighterTypeEnum::Cpp) {
+        QStringList lines = toPlainText().split("\n");
```

Comment:
Ok maybe not so common, but one may write multiple lines on thre same line
maybe should also split on that somehow, 

double var; int othervar;

But difficult since for (a; b; c) also contains mutliple ; on same line (but inside parenthesis ()

Reply:
True, but I think we can survive without that feature at the moment :)